kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -678,61 +698,90 @@ impl Socket for StreamSocket {
-                options.tcp.set_congestion(*congestion);
-            },
-            tcp_maxseg: MaxSegment => {
-                const MIN_MAXSEG: u32 = 536;
-                const MAX_MAXSEG: u32 = 65535;
-
-                let maxseg = tcp_maxseg.get().unwrap();
-                if *maxseg < MIN_MAXSEG || *maxseg > MAX_MAXSEG {
-                    return_errno_with_message!(Errno::EINVAL, "the maximum segment size is out of bounds");
-                }
-                options.tcp.set_maxseg(*maxseg);
-            },
-            tcp_window_clamp: WindowClamp => {
-                let window_clamp = tcp_window_clamp.get().unwrap();
-                let half_recv_buf = options.socket.recv_buf() / 2;
-                if *window_clamp <= half_recv_buf {
-                    options.tcp.set_window_clamp(half_recv_buf);
-                } else {
-                    options.tcp.set_window_clamp(*window_clamp);
-                }
-            },
-            _ => return_errno_with_message!(Errno::ENOPROTOOPT, "the socket option to be set is unknown")
-        });
-
         Ok(())
     }
 }
 
+fn do_tcp_setsockopt(
+    option: &dyn SocketOption,
+    options: &mut OptionSet,
+    state: &mut State,
+) -> Result<NeedIfacePoll> {
+    match_sock_option_ref!(option, {
+        tcp_no_delay: NoDelay => {
+            let no_delay = tcp_no_delay.get().unwrap();
+            options.tcp.set_no_delay(*no_delay);
+            state.set_raw_option(|raw_socket: &mut dyn RawTcpSetOption| raw_socket.set_nagle_enabled(!no_delay));
+        },
+        tcp_maxseg: MaxSegment => {
+            const MIN_MAXSEG: u32 = 536;
+            const MAX_MAXSEG: u32 = 65535;
+
+            let maxseg = tcp_maxseg.get().unwrap();
+            if *maxseg < MIN_MAXSEG || *maxseg > MAX_MAXSEG {
+                return_errno_with_message!(Errno::EINVAL, "the maximum segment size is out of bounds");
+            }
+            options.tcp.set_maxseg(*maxseg);
+        },
+        tcp_keep_idle: KeepIdle => {
+            const MIN_KEEP_IDLE: u32 = 1;
+            const MAX_KEEP_IDLE: u32 = 32767;
+
+            let keepidle = tcp_keep_idle.get().unwrap();
+            if *keepidle < MIN_KEEP_IDLE || *keepidle > MAX_KEEP_IDLE {
+                return_errno_with_message!(Errno::EINVAL, "the keep idle time is outside of bounds");
+            }
+            let duration = {
+                let now = Jiffies::elapsed().as_duration();
+                now + Duration::from_secs(*keepidle as _)
+            };
+            options.tcp.set_keep_idle(Some(duration));
+
+            // FIXME: Initiate a timer to begin the keep-alive probe after a specified duration.
+            //
+            // 'Keepidle' refers to the requirement for the connection to remain idle
+            // for a certain period before TCP begins sending keep-alive probes.
+            // For simplicity, we are configuring the socket keep-alive immediately.
+            state.set_keep_alive(true);
```

Comment:
>        TCP_KEEPIDLE (since Linux 2.4)
>              The time (in seconds) the connection needs to remain idle
>              before TCP starts sending keepalive probes, if the socket
>              option SO_KEEPALIVE has been set on this socket.  This
>              option should not be used in code intended to be portable.

Note that: **if the socket option SO_KEEPALIVE has been set on this socket**

So I guess we don't need this `set_keep_alive(true)`. If `SO_KEEPALIVE` is not set, `TCP_KEEPIDLE` will have no effect at all.

Reply:
Thanks for your points! I think I misunderstood the meaning of this option. Now it have been fixed.